monolithic_integration_test/build.rs

```diff
@@ -2,75 +2,163 @@ fn main() {
-        let build_output = Command::new("docker")
-            .args([
-                "build",
-                "--file=daphne_worker_test/docker/miniflare.Dockerfile",
-                "--quiet",
-                ".",
-            ])
-            .current_dir(&daphne_checkout_dir)
-            .output()
-            .expect("Failed to execute `docker build` for test Daphne");
-        assert!(
-            build_output.status.success(),
-            "Docker build of test Daphne failed:\n{}",
-            String::from_utf8_lossy(&build_output.stderr)
-        );
-        let image_id = String::from_utf8(build_output.stdout).unwrap();
-        let image_id = image_id.trim();
+        if container_strategy == "build" {
+            // Check out Daphne repository at a fixed hash, then build & save off a container image for
+            // the test Daphne instance.
+            const DAPHNE_COMMIT_HASH: &str = "e1b503eb2aefadfe2717abb0a359892848175534";
+            let daphne_checkout_dir = tempdir().unwrap();
+            let clone_output = Command::new("git")
+                .args(["clone", "-n", "https://github.com/cloudflare/daphne", "."])
+                .current_dir(&daphne_checkout_dir)
+                .output()
+                .expect("Failed to execute `git clone` for Daphne repository");
+            assert!(
+                clone_output.status.success(),
+                "Git clone of Daphne repository failed:\n{}",
+                String::from_utf8_lossy(&clone_output.stderr)
+            );
+            let checkout_output = Command::new("git")
+                .args(["checkout", DAPHNE_COMMIT_HASH])
+                .current_dir(&daphne_checkout_dir)
+                .output()
+                .expect("Failed to execute `git checkout` for Daphne repository");
+            assert!(
+                checkout_output.status.success(),
+                "Git checkout of Daphne repository failed:\n{}",
+                String::from_utf8_lossy(&checkout_output.stderr)
+            );
 
-        let image_file = File::create(format!(
-            "{}/test_daphne.tar.zst",
-            env::var("OUT_DIR").unwrap()
-        ))
-        .expect("Couldn't create test Daphne image file");
-        save_zstd_compressed_docker_image(image_id, &image_file);
-        image_file
-            .sync_all()
-            .expect("Couldn't write compressed image file");
-        drop(image_file);
+            // Note: `docker build` has an `--output` flag which writes the output to somewhere, which
+            // may be a tarfile. But `docker build --output` only exports the image filesystem, and not
+            // any other image metadata (such as exposed ports, the entrypoint, etc), so we can't easily
+            // use it.
+            let build_output = Command::new("docker")
+                .args([
+                    "build",
+                    "--file=daphne_worker_test/docker/miniflare.Dockerfile",
+                    "--quiet",
```

Comment:
This is going to break if the layout of the Daphne repository changes, right?

Reply:
Yes, we will have to update these instructions to account for backwards-incompatible changes in Daphne's build system when we bump Daphne versions, similarly to the human-targeted instructions. (note that we fix a version of Daphne under test, so changes to the Daphne repo won't trigger the need to make these changes automatically)